esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,286 @@
+# Key-Value Store Intrinsics
+
+## Simple Summary
+
+This ESR defines a potential set of intrinsics to implement part of [Key-Value Database](esr_key_value_database.md).
+It doesn't cover CDT enhancements.
+
+## Database ID
+
+The `db` argument indicates which database to operate on. There are initially 2 values for this:
+
+* `"eosio.kvram"`: This database has low latency and consumes `RAM` resources.
+* `"eosio.kvdisk"`: This database consumes a new `DISK` resource. It may have a higher capacity, but may also consume higher
+       CPU to access than database `"eosio.kvram"`.
+
+All intrinsics which have a `db` argument assert if it's out of range.
+
+## Temporary Data Buffer
+
+Some intrinsics use a temporary data buffer. `kv_get_data` reads this buffer. The buffer is initially empty.
+
+## Ordering
+
+Keys are ordered lexicographically by `uint8_t`.
+
+## Non-Iterator Intrinsics
+
+```c++
+void     kv_erase(uint64_t db, uint64_t contract, const char* key, uint32_t key_size);
+void     kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+bool     kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+```
+
+These intrinsics modify the database and support point lookups. Point lookups may use less CPU than iterator operations.
+
+### kv_erase
+
+`kv_erase` erases a key. If `contract` doesn't match the executing contract then it aborts the transaction. Else the
+function succeeds, even if the key didn't exist.
+
+If any iterators are positioned at the erased key, then `kv_erase` switches their status to `iterator_erased`.
+
+`kv_erase` clears the temporary data buffer.
+
+### kv_set
+
+`kv_set` sets a key-value pair. If `contract` doesn't match the executing contract then it aborts the transaction.
+If the contract exceeds available resources, it will fail at the appropriate time, similar to existing database
+primitives.
```

Comment:
Resources are always billed to the contract, right?

Reply:
yes